### PR TITLE
Generalize helper methods

### LINKED ISSUE #172: Abstract/parameterize helper methods with lambdas
`ByName.getAllValues()` and `ByToken.getAllValues()` only differ in their comparison code to determine a match. These methods can be merged and parameterized with a lambda that implements the comparison.

Additionally, inspect other helper code for similar cases and resolve those too.

### PR DESCRIPTION
**Pull request by mvanaken**
Resolves #172.

### CODE REVIEW

File: core/src/main/java/io/parsingdata/metal/data/transformation/Wrapping.java [hunk 1]
```diff
@@ -30,14 +30,14 @@
 
     private Wrapping() {}
 
-    public static <T> ImmutableList<Optional<T>> wrap(final ImmutableList<T> list) {
+    public static <T, U extends T> ImmutableList<Optional<T>> wrapToBase(final ImmutableList<U> list) {
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/transformation/Wrapping.java]: I would be OK if we just keep the name `wrap()`.

File: core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java]: It's 2017 😉

File: core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import io.parsingdata.metal.data.ParseValue;
+
+public final class ByPredicate {
+
+    private ByPredicate() {}
+
+    public static final int NO_LIMIT = -1;
+
+    static SafeTrampoline<ImmutableList<ParseValue>> getAllValues(final ImmutableList<ParseGraph> graphList, final ImmutableList<ParseValue> valueList, final Predicate<ParseValue> predicate, final int limit) {
+        if (graphList.isEmpty() || valueList.size == limit) { return complete(() -> valueList); }
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java]: I think we have tests for "get all" and "get last", but no explicit test for something like "get the last 3". Maybe add that?

Also: this is not (yet?) exposed to the `ref()` and `token()` expressions, right? Is there an issue yet to add that?

File: core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java (hunk 2 above)
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java]: Agreed. There's a separate issue however, because it's best if it's one big commit to fix this: #133.

File: core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java (hunk 3 above)
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByPredicate.java]: #169. And good point about explicitly testing other values for `limit` (possibly also weird edge cases, where `limit` > available, etc.

However, I have some additional refactoring plans that may make those tests unnecessary, so my proposal is to not dive into adding many tests. But just so we don't forget: #180.
